OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs

```diff
@@ -0,0 +1,198 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Deliver produced units from outside the map.")]
+	public class ProductionByDeliveryInfo : ProductionInfo
+	{
+		[Desc("Delivering actor name.")]
+		[ActorReference]
+		public readonly string DeliveryActor = "c17";
+
+		public readonly string ReadyAudio = "Reinforce";
+
+		public override object Create(ActorInitializer init) { return new ProductionByDelivery(init, this); }
+	}
+
+	class ProductionByDelivery : Production
+	{
+		readonly CPos startPos;
+		readonly CPos endPos;
+		readonly ProductionByDeliveryInfo info;
+
+		Actor self;
+		string factionVariant;
+
+		public ProductionByDelivery(ActorInitializer init, ProductionByDeliveryInfo info)
+			: base(init, info)
+		{
+			self = init.Self;
+
+			// Start a fixed distance away: the width of the map.
+			// This makes the production timing independent of spawnpoint
+			startPos = self.Location + new CVec(self.Owner.World.Map.Bounds.Width, 0);
+			endPos = new CPos(self.Owner.World.Map.Bounds.Left - 5, self.Location.Y);
+			this.info = info;
+		}
+
+		public override Actor DoProduction(Actor self, ActorInfo actorToProduce, ExitInfo exitInfo, string raceVariant)
+		{
+			var exit = self.Location + exitInfo.ExitCell;
+			var spawn = self.CenterPosition + exitInfo.SpawnOffset;
+
+			var td = new TypeDictionary
+			{
+				new OwnerInit(self.Owner),
+				new LocationInit(exit),
+				new CenterPositionInit(spawn)
+			};
+
+			if (!string.IsNullOrEmpty(raceVariant))
+				td.Add(new RaceInit(raceVariant));
+
+			return self.World.CreateActor(false, actorToProduce.Name, td);
+		}
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			var deliveringActorType = info.DeliveryActor;
+			var owner = self.Owner;
+
+			this.self = self;
+			factionVariant = raceVariant;
+
+			// Check if there is a valid drop-off point before sending the transport
+			var exit = GetAvailableExit(self, actorsToProduce.First());
+			if (exit == null)
+				return false;
+
+			foreach (var trait in self.TraitsImplementing<INotifyDelivery>())
+				trait.IncomingDelivery(self);
+
+			owner.World.AddFrameEndTask(w =>
+			{
+				var altitude = self.World.Map.Rules.Actors[deliveringActorType].Traits.Get<PlaneInfo>().CruiseAltitude;
```

Comment:
How about `AircraftInfo` instead? Then people can use helicopters for delivery as well.